ops/pebble.py

```diff
@@ -1759,20 +1998,23 @@ def _make_auth_dict(permissions, user_id, user, group_id, group) -> typing.Dict:
             d['group'] = group
         return d
 
-    def _encode_multipart(self, metadata, path, source, encoding):
+    def _encode_multipart(self, metadata: Dict[str, Any], path: str,
+                          source: '_IOSource', encoding: str):
         # Python's stdlib mime/multipart handling is screwy and doesn't handle
         # binary properly, so roll our own.
-
-        if isinstance(source, str):
-            source = io.StringIO(source)
-        elif isinstance(source, bytes):
-            source = io.BytesIO(source)
-
+        def _get_source_io() -> '_AnyStrFileLikeIO':
+            if isinstance(source, str):
+                return io.StringIO(source)  # type: ignore  # protocol unhappy
+            elif isinstance(source, bytes):
+                return io.BytesIO(source)  # type: ignore  # protocol unhappy
+            return source
+
+        source_io = _get_source_io()
```

Comment:
This seems to work well:

```python
        # Python's stdlib mime/multipart handling is screwy and doesn't handle
        # binary properly, so roll our own.
        if isinstance(source, str):
            source_io = io.StringIO(source)  # type: _AnyStrFileLikeIO
        elif isinstance(source, bytes):
            source_io = io.BytesIO(source)  # type: _AnyStrFileLikeIO
        else:
            source_io = source  # type: _AnyStrFileLikeIO

```

Pyright picks up on the isinstance constraining of source.

Reply:
ha, another case where pyright and pycharm's own type checker disagree. I was trying to satisfy both, but your code reads better.